### PR TITLE
Introduce `transform` crate

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR moves transformation logic out of the `expr` crate and in to its own crate. The rationale is that `expr` is a very popular crate, but many of its users do not depend on the state of transformations. In particular, the `dataflow` crate is expensive to rebuild and while it does depend on `expr` it does not depend on transformations.

This PR also removes three transformations that are no longer in use, and flags two for further review. It also slaps up documentation everywhere. I'd love to require doc code examples, but the lint (`missing-doc-code-examples`) doesn't seem to have an effect that I could tell.

To summarize the commits: 

1. Move `src/expr/transform` to its own crate, introduce and correct references.
2. Remove three dead transformations.
3. Add documentation all over the place.
4. `cargo fmt`, obviously.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/2911)

### CODE REVIEW

File: src/transform/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+num_enum = "0.4.2"
+ordered-float = { version = "1.0.2", features = ["serde"] }
+ore = { path = "../ore" }
+pdqselect = "0.1.0"
+regex = "1.3.7"
+regex-syntax = "0.6.15"
+repr = { path = "../repr" }
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
+unicase = "2.6.0"
```
Comment (benesch) [inline src/transform/Cargo.toml:31]: I would suspect that you can ditch almost all of these dependencies?

**Comment by benesch** [comment]
Oh, I didn't review the documentation in any great detail. Perhaps that's something that @justinj would be interested in? But also doesn't seem like it needs to block merging.

**Comment by frankmcsherry** [comment]
>  I wonder if we should call this expr-transform to be a bit less generic? But I don't really have a ton of feelings on the matter.

I missed this somehow, sorry. I'm up for this maybe, though one thing that did move into here was the one `DataflowDesc` transformation. Probably more coming. But up for a smarter name.

**Comment by benesch** [comment]
> > I wonder if we should call this expr-transform to be a bit less generic? But I don't really have a ton of feelings on the matter.
> 
> I missed this somehow, sorry. I'm up for this maybe, though one thing that did move into here was the one `DataflowDesc` transformation. Probably more coming. But up for a smarter name.

`dataflow-transform` perhaps? Real easy to do later, though, especially considering that there is only one thing that depends on this crate.
